File: javascript/enigma.js
import { Rotor } from "./rotor";
import { Reflecteur } from "./reflecteur";

class Enigma {
    constructor(rotors, reflecteur) {
        this.rotors = new Array();
        this.rotor_rotation = new Array();

        this.N = rotors[0].length;
        this.N_rotors = rotors.length;

        for(let i = 0; i < this.N_rotors; i++) {
            if (rotors[i].length != this.N) {
                alert("Les rotors doivent faire la même taille !");
                throw new Error("Les rotors doivent faire la même taille !");
            }
            this.rotors.push(new Rotor(rotors[i]));
            this.rotor_rotation.push(0);
        }

        if( reflecteur.length != this.N) {
            alert("Les rotors  et le reflecteur doivent faire la même taille !");
            throw new Error("Les rotors  et le reflecteur doivent faire la même taille !");
        }

        this.reflecteur = new Reflecteur(reflecteur);
    }

    tourner() {
        this.rotors[0].tourner();
        this.rotor_rotation[0]++;
        
        let i = 0;
        while ((this.rotor_rotation[i] % this.N == 0) && (i != self.N_rotors)) {
            this.rotor_rotation[i] = 0;
            this.rotors[i+1].tourner();
            this.rotor_rotation[i+1]++;
            i++;
        }
    }

    coder(text) {
        let texte_code = "";
        text = text.toUpperCase();

        for (let i = 0; i < text.length; i++) {
            let lettre_codee = text.charCodeAt(i) - 65;
            if(lettre_codee >= 0 && lettre_codee <= 25) {
                for(let j = 0; j < this.N_rotors; j++) {
                    lettre_codee = this.rotors[j].coder(lettre_codee);
                }

                lettre_codee = this.reflecteur.coder(lettre_codee);

                for(let j = this.N_rotors - 1; j >= 0; j--) {
                    lettre_codee = this.rotors[j].decoder(lettre_codee);
                }

                texte_code += String.fromCharCode(65 + lettre_codee);

                this.tourner();
            }
            else {
                texte_code += String.fromCharCode(65 + lettre_codee);
            }
        }

        return texte_code;
    }

    position_initiale() {
        for(let i = 0; i < this.N_rotors; i++) {
            for( let j = 0; j < (this.N - this.rotor_rotation[i]) % this.N; j++) {
                this.rotors[i].tourner();
            }
            this.rotor_rotation[i] = 0;
        }
    }
}

export { Enigma, Rotor, Reflecteur };

File: python/enigma.py
from rotor import *
from reflecteur import *

class Enigma:
    def __init__(self, rotors, reflecteur):
        self.__rotors = []

        self.__N = len(rotors[0])

        # On ajoute les rotors à notre machine et on vérifie qu'ils ont tous la même taille
        for rotor in rotors:
            if len(rotor) != self.__N:
                raise Exception("Les rotors doivent être de même taille")
            self.__rotors.append(Rotor(rotor))

        # Nombre de rotors
        self.__N_rotors = len(self.__rotors)

        # Position des rotors (0 = il n'a pas bougé, 1 = il a tourné 1 fois ..)
        self.__rotor_rotation = [0] * self.__N_rotors

        # On ajoute le reflecteur après avoir vérifier qu'il avait la bonne taille
        if len(reflecteur) != self.__N:
            raise Exception("Le reflecteur doit être de la même taille que les rotors")

        self.__reflecteur = Reflecteur(reflecteur)

    # Fonction pour coder un texte avec notre machine enigma
    def coder(self, text):
        text_code = ""

        for lettre in text:
            # Si la le caractère est une lettre de l'alphabet, on le code :
            if lettre.upper() in alphabet:
                lettre_codee = alphabet.index(lettre.upper())

                # On code la lettre à travers tous les rotors
                for rotor in self.__rotors:
                    lettre_codee = rotor.coder_nombre(lettre_codee)

                # Puis le reflecteur
                lettre_codee = self.__reflecteur.coder_nombre(lettre_codee)

                # Et enfin à travers les rotors mais dans l'autre sens
                for rotor in reversed(self.__rotors):
                    lettre_codee = rotor.decoder_nombre(lettre_codee)

                text_code += alphabet[lettre_codee]

                # On fait tourner les rotors
                self.__tourner()
            # Sinon, on ne le code pas :
            else:
                text_code += lettre

        return text_code

    # Fait tourner les rotors
    def __tourner(self):
        # Le premier rotor tourne tout le temps        
        self.__rotors[0].tourner()
        self.__rotor_rotation[0] += 1

        # Si le rotor a fait un nombre de tours divisible par le nombre de lettres par rotor, on fait tourner celui d'après
        i = 0
        while (self.__rotor_rotation[i] % self.__N == 0) and (i < self.__N_rotors - 1):
            self.__rotor_rotation[i] = 0
            self.__rotors[i+1].tourner()
            self.__rotor_rotation[i+1] += 1
            i += 1

    # Remet la machine enigma dans sa position initiale
    def position_initiale(self):
        # Pour chaque rotor :
        for i in range(self.__N_rotors):
            # On le fait tourner assez de fois pour qu'il retrouve sa position initiale :
            for j in range((self.__N - self.__rotor_rotation[i]) % self.__N):
                self.__rotors[i].tourner()
            self.__rotor_rotation[i] = 0